storage/imx6ull-flash/flashsrv.c

```diff
@@ -114,6 +124,82 @@ static int flashsrv_devfsSymlink(const char *name, const char *target)
 }
 
 
+/* Partition table access functions (ptable buffer must have mtd->writesz size) */
+
+static int flashsrv_ptableRead(storage_t *strg, ptable_t *ptable)
+{
+	storage_mtd_t *mtd = strg->dev->mtd;
+	size_t retlen;
+	off_t offs;
+	int err, i;
+
+	/* Find first valid partition table copy */
+	for (i = 0; i < PTABLE_NCOPIES; i++) {
+		offs = strg->size - (i + 1) * mtd->erasesz;
+		if (mtd->ops->block_isBad(strg, offs)) {
+			continue;
+		}
+
+		err = mtd->ops->read(strg, offs, ptable, mtd->writesz, &retlen);
+		/* -EUCLEAN isn't a fatal error (indicates dangerous page degradation but all bitflips were successfully corrected) */
+		if (((err < 0) && (err != -EUCLEAN)) || (retlen != mtd->writesz)) {
+			continue;
+		}
+
+		/* Partition table must fit in a NAND page */
+		if (ptable_size(ptable->count) > mtd->writesz) {
+			continue;
+		}
+
+		if (ptable_deserialize(ptable, strg->size, strg->dev->mtd->erasesz) < 0) {
+			continue;
+		}
+
+		return EOK;
+	}
+
+	return -ENOENT;
+}
+
+
+static int flashsrv_ptableWrite(storage_t *strg, ptable_t *ptable)
+{
+	storage_mtd_t *mtd = strg->dev->mtd;
+	size_t retlen;
+	off_t offs;
+	int err, i, n = 0;
+
+	/* Partition table must fit in a NAND page */
+	if (ptable_size(ptable->count) > mtd->writesz) {
+		return -EINVAL;
+	}
+
```

Comment:
Hmmm, maybe we should check if ptable overlaps the last partition?

Reply:
Not sure if we can do this because we currently have 2 approaches:
1. There is no separate partition created for ptable (in this case we can find last partition and make the check).
2. There is a partition added for partition table (in this case `ptable_verify()` will make sure that there is no overlap with other partitions).

The chosen approach depends on the project and we don't know which one is in use.

Reply:
Ack